File: frontend/src/shared/assets/spriteActions.ts
// Mirrors backend/assets/sprite_actions.py. Named sprite slots the monster
// upload form exposes; frames are counted server-side from each sheet.
export const MONSTER_ACTIONS = [
  'idle',
  'walk',
  'attack',
  'hurt',
  'death',
  'portrait',
  'projectile',
] as const

export type MonsterAction = (typeof MONSTER_ACTIONS)[number]

/** Minimum sprites a monster needs to stand and land a hit in battle. */
export const REQUIRED_MONSTER_ACTIONS: MonsterAction[] = ['idle', 'attack']

/** Actions that loop while displayed; the rest play once. */
export const LOOPING_ACTIONS: MonsterAction[] = ['idle', 'walk', 'projectile']

/** Suggested default fps per action, used to prefill the upload form. */
export const DEFAULT_ACTION_FPS: Record<MonsterAction, number> = {
  idle: 8,
  walk: 10,
  attack: 11,
  hurt: 10,
  death: 8,
  portrait: 1,
  projectile: 14,
}

File: backend/assets/sprite_actions.py
"""Sprite action categories shared across seed data, upload, and the frontend.

The ``AssetSprite.action`` field is a free-form slug, but the game only renders a
known set. Centralising them here keeps the monster-upload form, validator, and
seed importer in agreement. The frontend mirrors this in
``frontend/src/shared/assets/spriteActions.ts``.
"""

from __future__ import annotations

# Ordered for display in the upload form. ``projectile`` only applies to ranged
# attackers (``attack.kind == "projectile"``).
MONSTER_ACTIONS: tuple[str, ...] = (
    "idle",
    "walk",
    "attack",
    "hurt",
    "death",
    "portrait",
    "projectile",
)

# The minimum a monster needs to stand in a battle and land a hit.
REQUIRED_MONSTER_ACTIONS: frozenset[str] = frozenset({"idle", "attack"})

# Actions that loop while displayed; everything else plays once.
LOOPING_ACTIONS: frozenset[str] = frozenset({"idle", "walk", "projectile"})
